[ROCm][CI] Build theRock with pytorch nightly

ethanwee1 opened the pull request:
This PR aims to build  the PyTorch nightly branch against theRock nightly ROCm packages. 
A new `rocm-nightly.yml` workflow was added to run CI builds targeting gfx942 (MI300). In the future we will add MI355 support. TheRock installs via wheels, so the traditional `/opt/rocm/ `installation path is no longer valid. We will have a followup PR to enable testing with TheRock https://github.com/pytorch/pytorch/pull/170105
 
NOTE: PyTorch is built *without* Magma support for nightly ROCm as of this PR.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang @EikanWang @jgong5 @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela @xmfan @dllehr-amd

jithunnair-amd commented on .ci/docker/ubuntu-rocm/entrypoint.sh:
```diff
@@ -0,0 +1,66 @@
+            # Export environment variables
+            export ROCM_PATH="$ROCM_PATH"
+            export ROCM_HOME="$ROCM_PATH"
+            export PATH="${ROCM_BIN}:${ROCM_PATH}/bin:${ROCM_PATH}/hcc/bin:${ROCM_PATH}/hip/bin:${ROCM_PATH}/opencl/bin:${ROCM_PATH}/llvm/bin:${PATH}"
```
As discussed, `hcc/bin`, `hip/bin`, `opencl/bin` might be long obsolete.

powderluv commented on .ci/docker/build.sh:
```diff
@@ -189,6 +189,18 @@ case "$tag" in
+    KATEX=yes
+    UCX_COMMIT=${_UCX_COMMIT}
+    UCC_COMMIT=${_UCC_COMMIT}
+    PYTORCH_ROCM_ARCH="gfx942"
```
its ok for now but we should plan for gfx950

powderluv commented on .ci/docker/common/install_cache.sh:
```diff
@@ -116,23 +119,23 @@ if [ -n "$ROCM_VERSION" ]; then
 
-  if [[ -e "/opt/rocm/hcc/bin/hcc" ]]; then
+  if [[ -e "${ROCM_PATH}/hcc/bin/hcc" ]]; then
     # ROCm 3.3 or earlier.
```
do we still care about this < ROCm 3.3?

jithunnair-amd replied:
Nope, we should drop it ... @ethanwee1

jithunnair-amd commented on .ci/docker/common/install_ucc.sh:
```diff
@@ -60,6 +66,7 @@ function install_ucc() {
     for arch in $amdgpu_targets; do
       HIP_OFFLOAD="$HIP_OFFLOAD --offload-arch=$arch"
     done
+    HIP_OFFLOAD="$HIP_OFFLOAD --rocm-path=${ROCM_PATH} --rocm-device-lib-path=${ROCM_PATH}/lib/llvm/amdgcn/bitcode"
```
Let's define `ROCM_DEVICE_LIB_PATH` in `detect_rocm_path.sh` and use it here, since the actual path depends on whether it's a legacy installation of ROCm or not

ethanwee1 committed:
[ROCm][CI] Add theRock nightly ROCm support

- Add detect_rocm_path.sh helper for dynamic ROCm path detection
- Update install_rocm.sh for theRock nightly pip-based installation
- Add entrypoint.sh for runtime ROCm path configuration
- Skip sccache wrapping for theRock nightly builds
- Add LD_LIBRARY_PATH for rocm_sysdeps (liblzma linking fix)
- Add rocm-nightly.yml workflow for CI testing
- Add HIP_DEVICE_LIB_PATH to cmake for device library discovery
- Update Dockerfile and build scripts for theRock compatibility

Tested: Docker image builds and PyTorch compiles successfully with gfx942

ethanwee1 committed:
Disable FBGEMM_GENAI for theRock nightly builds

ethanwee1 committed:
Restore USE_FBGEMM=0 for CUDA 13 builds

ethanwee1 committed:
Add roctracer include path for kineto on theRock nightly

ethanwee1 committed:
Set ROCM_SOURCE_DIR for Kineto to find roctracer headers in theRock nightly

ethanwee1 committed:
update for ROCM_SOURCE_DIR

ethanwee1 committed:
Fix shellcheck lint for rocm_env.sh source

ethanwee1 committed:
Skip clang restore for theRock nightly (sccache disabled, different LLVM path)

ethanwee1 committed:
Source ROCm environment in test.sh for theRock nightly builds

This ensures LD_LIBRARY_PATH is set correctly so Python can find
librocm_sysdeps_liblzma.so.5 when running tests.

ethanwee1 committed:
Remove entrypoint.sh from ubuntu-rocm Docker image

The entrypoint script was causing issues with the Docker build and is not
needed for TheRock nightly ROCm builds. The environment setup can be done
directly in the Dockerfile or at runtime.

ethanwee1 committed:
Remove test job from rocm-nightly workflow (build-only PR)

Keep only the build job for theRock nightly ROCm. Testing will be added
in a separate PR.

jeffdaily commented:
@pytorchbot merge -f "updates only rocm docker files; builds passing"

jeffdaily committed:
undo accidental removal of Docker images for CUDA and benchmarks

jeffdaily committed:
Update rocm-nightly.yml to remove temporary tag triggers

Removed temporary tag triggers from nightly workflow.

jithunnair-amd committed:
lint

jithunnair-amd committed:
lint

jithunnair-amd committed:
Lint

ethanwee1 committed:
Remove ROCM_DEVICE_LIB_PATH

jeffdaily committed:
lint

ethanwee1 committed:
Fix UCC build with TheRock nightly ROCm by setting device library path

TheRock installs bitcode files at lib/llvm/amdgcn/bitcode instead of
the standard amdgcn/bitcode location. Add ROCM_DEVICE_LIB_PATH to
detect_rocm_path.sh and use it in install_ucc.sh to pass
--rocm-device-lib-path to the compiler.

jithunnair-amd commented on .ci/pytorch/build.sh:
```diff
@@ -6,6 +6,12 @@ set -ex -o pipefail
+  # shellcheck disable=SC1091
+  source /etc/rocm_env.sh
+fi
+
```
Any sourcing of script to set required ROCm env vars should be done in `common.sh` since both `build.sh` and `test.sh` source it.

ethanwee1 committed:
Unify ROCm env vars into install_rocm.sh and /etc/rocm_env.sh

- install_rocm.sh now writes /etc/rocm_env.sh for both nightly and traditional ROCm
- install_rocm.sh appends 'source /etc/rocm_env.sh' to /etc/bash.bashrc
- detect_rocm_path.sh simplified to just source /etc/rocm_env.sh
- Removed redundant bashrc logic from Dockerfile
- All ROCm env vars (ROCM_PATH, ROCM_HOME, ROCM_BIN, ROCM_CMAKE, ROCM_DEVICE_LIB_PATH, etc.) now in one place

ethanwee1 committed:
Remove detect_rocm_path.sh and simplify ROCm env sourcing

All scripts now directly source /etc/rocm_env.sh which is created by
install_rocm.sh. This removes redundant fallback logic and the need
for the detect_rocm_path.sh wrapper script.

ethanwee1 committed:
Move ROCm env sourcing to common.sh for build and test scripts

ethanwee1 committed:
Lint

ethanwee1 committed:
Fix install_ucc.sh to check for rocm_env.sh before sourcing

CUDA builds also use install_ucc.sh but don't have /etc/rocm_env.sh
since they don't run install_rocm.sh. Add existence check to prevent
build failures on CUDA images.

jithunnair-amd committed:
Clarify ROCm environment sourcing comment

Updated comment to clarify the purpose of sourcing ROCm environment.

jithunnair-amd reviewed (approved):
LGTM! Thanks for taking it through various rounds of refactor, @ethanwee1

jithunnair-amd commented:
@pytorchbot merge -f "CI was fully green in penultimate commit: https://hud.pytorch.org/pytorch/pytorch/pull/168377?sha=0c3bddffe48086ab5d0c0b7a740c6b4d7fa82e19, last commit was only a comment change, hence merging"
